coalib/misc/Caching.py

```diff
@@ -0,0 +1,168 @@
+import appdirs
+import calendar
+import os
+import pickle
+import time
+
+from coalib.misc.Constants import caching_db, changed_files_db
+
+
+def get_config_data_path(filename):
+    """
+    Get the full path of ``filename`` present in the user config directory.
+
+    :param filename: The file whose path needs to be expanded.
+    :return:         Full path of the file, assuming it's present in the
+                     user's config directory.
+    """
+    path = appdirs.user_config_dir("coala")
+    return os.path.join(path, filename)
+
+
+def pickle_load(filename, fallback=None):
+    """
+    Get the data stored in ``filename`` present in the coala user
+    config directory.
+
+    :param filename: The name of the file present in the coala user config
+                     directory.
+    :param fallback: Return value to fallback to in case the file doesn't
+                     exist.
+    :return:         Data that is present in the file, if the file exists.
+                     Otherwise the ``default`` value is returned.
+    """
+    filename = get_config_data_path(filename)
+    if not os.path.isfile(filename):
+        return fallback
+    with open(filename, "rb") as f:
+        return pickle.load(f)
+
+
+def pickle_dump(filename, data):
+    """
+    Write ``data`` into the file ``filename`` present in the coala user
+    config directory.
+
+    :param filename: The name of the file present in the coala user config
+                     directory.
+    :param data:     Data to be serialized and written to the file using
+                     pickle.
+    """
+    filename = get_config_data_path(filename)
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+    with open(filename, "wb") as f:
+        return pickle.dump(data, f)
+
+
+def verify_file_integrity(log_printer, files=[caching_db, changed_files_db]):
+    """
+    Verify that the pickled database files are not corrupted.
+
+    :return: True if the files are not corrupted.
+             False if the files are corrupted. Also, the files are deleted
+             if they are corrupted.
+    """
+    ret = True
+    for file_name in files:
+        file_path = get_config_data_path(file_name)
+        if not os.path.isfile(file_path):
+            continue
+        with open(file_path, "rb") as f:
+            try:
+                pickle.load(f)
+            except pickle.UnpicklingError:
+                ret = False
+    if not ret:
+        log_printer.warn("The caching database is corrupted and will "
+                         "be removed. Each project will be re-cached "
+                         "automatically the next time coala is run.")
+        for file_name in files:
+            file_path = get_config_data_path(file_name)
+            os.remove(file_path)
+
+    return ret
+
+
+def add_to_changed_files(project_dir, changed_files):
+    """
+    Save the set of files in ``changed_files`` to disk for future use
+    in ``update_last_coala_run_time``.
```

Comment:
maybe do a few doctests? I really think we should use them more often, cleaner tests and they're definitely not harmful to the docs.